### PR TITLE
Postgresql Provider is working excluding array types

### PR DESCRIPTION
**Pull request by WasimAhmad**

### CODE REVIEW

**Comment by Mimetis** [comment]
Thanks for the PR @WasimAhmad !!

Unfortunately, I guess you did not pull the master branch for a while because your provider is currently implementing an old version of **DMS** (I guess < `0.9.5`) and there are some changes that are not compatible now (And some Interfaces have been removed, resulting in compilation errors with your provider)

Can you make a PULL on the master branch and try to resolve the errors raised in your provider?
You can have a look on the other providers to see what the new format of the scope tables are.

Basically, scopes tables are now the same on server and on client, with some information moved from scope table to scope client table.

I'm still watching at your PR during that time !

Once again, it's an awesome job you just did here !

**Comment by WasimAhmad** [comment]
my fault. 
Forgot to remove old files. fixed now

**Comment by WasimAhmad** [comment]
should I create the PR again?

**Comment by Mimetis** [comment]
Nop, just update your branch

**Comment by WasimAhmad** [comment]
its done

**Comment by Mimetis** [comment]
I'm trying to sync the AdventureWorks sample database, from **SQL Server** to **Postgres** but I have a problem when trying to sync a table with a **DateTime** column (I'm trying with `ProductCategory` table)

image

**Comment by WasimAhmad** [comment]
ok let me check

**Comment by Mimetis** [comment]
I have some others problem when the table has multiple primary keys (for example `CustomerAddress`)

I think a good test you can do, is to try to sync the [AdventureWorks database](https://github.com/Mimetis/Dotmim.Sync/blob/master/CreateAdventureWorks.sql) from **SQL Server** to **PostgreSQL**.

And specifying these tables:

``` cs
SqlSyncProvider serverProvider = new SqlSyncProvider("Data Source=.....");
NpgsqlSyncProvider clientProvider = new NpgsqlSyncProvider("Host=...");

// Tables involved in the sync process:
var setup = new SyncSetup("ProductCategory", "ProductDescription", "ProductModel", 
                          "Product", "ProductModelProductDescription", "Address", 
                          "Customer", "CustomerAddress", "SalesOrderHeader", "SalesOrderDetail");

// Sync agent
SyncAgent agent = new SyncAgent(clientProvider, serverProvider);

do
{
    var result = await agent.SynchronizeAsync(setup);
    Console.WriteLine(result);

} while (Console.ReadKey().Key != ConsoleKey.Escape);
```

**Comment by WasimAhmad** [comment]
> I'm trying to sync the AdventureWorks sample database, from **SQL Server** to **Postgres** but I have a problem when trying to sync a table with a **DateTime** column (I'm trying with `ProductCategory` table)
> 
> image

this is fixed.

**Comment by WasimAhmad** [comment]
> I have some others problem when the table has multiple primary keys (for example `CustomerAddress`)
> 
> I think a good test you can do, is to try to sync the [AdventureWorks database](https://github.com/Mimetis/Dotmim.Sync/blob/master/CreateAdventureWorks.sql) from **SQL Server** to **PostgreSQL**.
> 
> And specifying these tables:
> 
> ```cs
> SqlSyncProvider serverProvider = new SqlSyncProvider("Data Source=.....");
> NpgsqlSyncProvider clientProvider = new NpgsqlSyncProvider("Host=...");
> 
> // Tables involved in the sync process:
> var setup = new SyncSetup("ProductCategory", "ProductDescription", "ProductModel", 
>                           "Product", "ProductModelProductDescription", "Address", 
>                           "Customer", "CustomerAddress", "SalesOrderHeader", "SalesOrderDetail");
> 
> // Sync agent
> SyncAgent agent = new SyncAgent(clientProvider, serverProvider);
> 
> do
> {
>     var result = await agent.SynchronizeAsync(setup);
>     Console.WriteLine(result);
> 
> } while (Console.ReadKey().Key != ConsoleKey.Escape);
> ```

ok i will get into it

**Comment by WasimAhmad** [comment]
composite primary key issue also fixed..
image

tested using 
 ```csharp
var setup = new SyncSetup("ProductCategory", "ProductDescription", "ProductModel", 
                          "Product", "ProductModelProductDescription", "Address", 
                          "Customer", "CustomerAddress", "SalesOrderHeader", "SalesOrderDetail");
``

**Comment by Mimetis** [comment]
Nice!!

Indeed, it's working great now

I'm still continuing to test your provider.
Now the problem is coming when we have a filter.

Here is my test:

``` cs
SqlSyncProvider serverProvider = new SqlSyncProvider("Data Source=.....");
NpgsqlSyncProvider clientProvider = new NpgsqlSyncProvider("Host=...");

// Tables involved in the sync process:
var setup = new SyncSetup("ProductCategory");
setup.Filters.Add("ProductCategory", "ParentProductCategoryID", null, true);

// Sync agent
SyncAgent agent = new SyncAgent(clientProvider, serverProvider);

do
{
    var result = await agent.SynchronizeAsync(setup);
    Console.WriteLine(result);

} while (Console.ReadKey().Key != ConsoleKey.Escape);
```


image

occuring in the `CreateFilterWhereSide` method from the `NpgsqlBuilderProcedure` class

Please take a look at the `MySqlSyncProvider`, I think it's most likely the best way to debug this

**Comment by WasimAhmad** [comment]
Ok let me check

**Comment by WasimAhmad** [comment]
> Nice!!
> 
> Indeed, it's working great now
> 
> I'm still continuing to test your provider. Now the problem is coming when we have a filter.
> 
> Here is my test:
> 
> ```cs
> SqlSyncProvider serverProvider = new SqlSyncProvider("Data Source=.....");
> NpgsqlSyncProvider clientProvider = new NpgsqlSyncProvider("Host=...");
> 
> // Tables involved in the sync process:
> var setup = new SyncSetup("ProductCategory");
> setup.Filters.Add("ProductCategory", "ParentProductCategoryID", null, true);
> 
> // Sync agent
> SyncAgent agent = new SyncAgent(clientProvider, serverProvider);
> 
> do
> {
>     var result = await agent.SynchronizeAsync(setup);
>     Console.WriteLine(result);
> 
> } while (Console.ReadKey().Key != ConsoleKey.Escape);
> ```
> 
> image
> 
> occuring in the `CreateFilterWhereSide` method from the `NpgsqlBuilderProcedure` class
> 
> Please take a look at the `MySqlSyncProvider`, I think it's most likely the best way to debug this

its fixed. i hope it does not break anything.. ;)

**Comment by WasimAhmad** [comment]
Any other problem?

**Comment by Mimetis** [comment]
Seems fine, but I have a lot of issues when trying to run tests.
I'm merging your PR in the master branch, but we need to pass all the tests to be able to publish a nuget package.

I've created a branch https://github.com/Mimetis/Dotmim.Sync/tree/AddNpgTests
This branch contains all the fixes I've done to be able to run a first test.

Can you look at it and see all the modification (by comparing your branch with this one)

The idea is to be able to firstly make all the tcp tests running. (look https://github.com/Mimetis/Dotmim.Sync/blob/AddNpgTests/Tests/Dotmim.Sync.Tests/IntegrationTests/Postgres/PostgresTcpTests.cs)

**Comment by Mimetis** [comment]
I've made a PR (draft mode) here https://github.com/Mimetis/Dotmim.Sync/pull/914

**Comment by WasimAhmad** [comment]
> Seems fine, but I have a lot of issues when trying to run tests. I'm merging your PR in the master branch, but we need to pass all the tests to be able to publish a nuget package.
> 
> I've created a branch https://github.com/Mimetis/Dotmim.Sync/tree/AddNpgTests This branch contains all the fixes I've done to be able to run a first test.
> 
> Can you look at it and see all the modification (by comparing your branch with this one)
> 
> The idea is to be able to firstly make all the tcp tests running. (look https://github.com/Mimetis/Dotmim.Sync/blob/AddNpgTests/Tests/Dotmim.Sync.Tests/IntegrationTests/Postgres/PostgresTcpTests.cs)

ok. I am looking at it.
